File: main/java/uk/co/real_logic/sbe/PrimitiveValue.java
/* -*- mode: java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil -*- */
/*
 * Copyright 2013 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe;

import static java.lang.Double.doubleToLongBits;

/**
 * Class used to encapsulate values for primitives. Used for nullValue, minValue, maxValue, and constants
 * <p/>
 * <table>
 *     <thead>
 *         <tr>
 *             <th>PrimitiveType</th>
 *             <th>Null</th>
 *             <th>Min</th>
 *             <th>Max</th>
 *         </tr>
 *     </thead>
 *     <tbody>
 *         <tr>
 *             <td>char</td>
 *             <td>0</td>
 *             <td>0x20</td>
 *             <td>0x7E</td>
 *         </tr>
 *         <tr>
 *             <td>int8</td>
 *             <td>-128</td>
 *             <td>-127</td>
 *             <td>127</td>
 *         </tr>
 *         <tr>
 *             <td>uint8</td>
 *             <td>255</td>
 *             <td>0</td>
 *             <td>254</td>
 *         </tr>
 *         <tr>
 *             <td>int16</td>
 *             <td>-32768</td>
 *             <td>-32767</td>
 *             <td>32767</td>
 *         </tr>
 *         <tr>
 *             <td>uint16</td>
 *             <td>65535</td>
 *             <td>0</td>
 *             <td>65534</td>
 *         </tr>
 *         <tr>
 *             <td>int32</td>
 *             <td>2^31</td>
 *             <td>-2^31 + 1</td>
 *             <td>2^31 - 1</td>
 *         </tr>
 *         <tr>
 *             <td>uint32</td>
 *             <td>2^32 - 1</td>
 *             <td>0</td>
 *             <td>2^32 - 2</td>
 *         </tr>
 *         <tr>
 *             <td>int64</td>
 *             <td>2^63</td>
 *             <td>-2^63 + 1</td>
 *             <td>2^63 - 1</td>
 *         </tr>
 *         <tr>
 *             <td>uint64</td>
 *             <td>2^64 - 1</td>
 *             <td>0</td>
 *             <td>2^64 - 2</td>
 *         </tr>
 *     </tbody>
 * </table>
 */
public class PrimitiveValue
{
    public enum Representation
    {
        LONG,
        DOUBLE
    }

    public static final long MIN_VALUE_CHAR = 0x20;
    public static final long MAX_VALUE_CHAR = 0x7E;
    public static final long NULL_VALUE_CHAR = 0;

    public static final long MIN_VALUE_INT8 = -127;
    public static final long MAX_VALUE_INT8 = 127;
    public static final long NULL_VALUE_INT8 = -128;

    public static final long MIN_VALUE_UINT8 = 0;
    public static final long MAX_VALUE_UINT8 = 254;
    public static final long NULL_VALUE_UINT8 = 255;

    public static final long MIN_VALUE_INT16 = -32767;
    public static final long MAX_VALUE_INT16 = 32767;
    public static final long NULL_VALUE_INT16 = -32768;

    public static final long MIN_VALUE_UINT16 = 0;
    public static final long MAX_VALUE_UINT16 = 65534;
    public static final long NULL_VALUE_UINT16 = 65535;

    public static final long MIN_VALUE_INT32 = -2147483647;
    public static final long MAX_VALUE_INT32 = 2147483647;
    public static final long NULL_VALUE_INT32 = -2147483648;

    public static final long MIN_VALUE_UINT32 = 0;
    public static final long MAX_VALUE_UINT32 = 0xFFFFFFFD;
    public static final long NULL_VALUE_UINT32 = 0xFFFFFFFE;

    public static final long MIN_VALUE_INT64 = Long.MIN_VALUE + 1;  // -2^63 + 1
    public static final long MAX_VALUE_INT64 = Long.MAX_VALUE;      //  2^63 - 1  (SBE spec says -2^63 - 1)
    public static final long NULL_VALUE_INT64 = Long.MIN_VALUE;     // -2^63

    public static final long MIN_VALUE_UINT64 = 0;
    public static final long MAX_VALUE_UINT64 = Long.MAX_VALUE;  // TODO: placeholder for now (replace with BigInteger?)
    public static final long NULL_VALUE_UINT64 = Long.MIN_VALUE; // TODO: placeholder for now (replace with BigInteger?)

    public static final float MIN_VALUE_FLOAT = Float.MIN_VALUE;
    public static final float MAX_VALUE_FLOAT = Float.MAX_VALUE;
    public static final float NULL_VALUE_FLOAT = Float.NaN;         // TODO: can NOT be used as a normal equality check

    public static final double MIN_VALUE_DOUBLE = Double.MIN_VALUE;
    public static final double MAX_VALUE_DOUBLE = Double.MAX_VALUE;
    public static final double NULL_VALUE_DOUBLE = Double.NaN;      // TODO: can NOT be used as a normal equality check

    private final Representation representation;
    private final long longValue;
    private final double doubleValue;

    /**
     * Construct and fill in value as a long.
     *
     * @param value in long format
     */
    public PrimitiveValue(final long value)
    {
        representation = Representation.LONG;
        longValue = value;
        doubleValue = 0.0;
    }

    /**
     * Construct and fill in value as a double.
     * @param value in double format
     */
    public PrimitiveValue(final double value)
    {
        representation = Representation.DOUBLE;
        longValue = 0;
        doubleValue = value;
    }

    /**
     * Parse constant value string and set representation based on type
     *
     *
     * @param value     expressed as a String
     * @param primitiveType that this is supposed to be
     * @return a new {@link PrimitiveValue} for the value.
     * @throws IllegalArgumentException if parsing not known for type
     */
    public static PrimitiveValue parse(final String value, final PrimitiveType primitiveType)
    {
        switch (primitiveType)
        {
            case CHAR:
                if (value.length() > 1)
                {
                    throw new IllegalArgumentException("constant char value malformed");
                }
                return new PrimitiveValue((long)value.getBytes()[0]);

            case INT8:
            case INT16:
            case INT32:
            case INT64:
            case UINT8:
            case UINT16:
            case UINT32:
            case UINT64:
                // TODO: not entirely adequate, but then again, Java doesn't have unsigned 64-bit integers...
                return new PrimitiveValue(Long.parseLong(value));

            case FLOAT:
            case DOUBLE:
                return new PrimitiveValue(Double.parseDouble(value));

            default:
                throw new IllegalArgumentException("Do not know how to parse this primitiveType type for constant value");
        }
    }

    /**
     * Return long value for this PrimitiveValue
     *
     * @return value expressed as a long
     * @throws IllegalArgumentException if not a long value representation
     */
    public long longValue()
    {
        if (representation != Representation.LONG)
        {
            throw new IllegalArgumentException("PrimitiveValue is not a long representation");
        }

        return longValue;
    }

    /**
     * Return string representation of this object
     *
     * @return String representing object value
     * @throws IllegalArgumentException if unknown representation
     */
    public String toString()
    {
        switch (representation)
        {
            case LONG:
                return Long.toString(longValue);

            case DOUBLE:
                return Double.toString(doubleValue);

            default:
                throw new IllegalStateException("Unsupported Representation: " + representation);
        }
    }

    /**
     * Determine if two values are equivalent.
     *
     * @param value to compare this value with
     * @return equivalence of values
     */
    public boolean equals(final Object value)
    {
        if (null != value && value instanceof PrimitiveValue)
        {
            PrimitiveValue rhs = (PrimitiveValue)value;

            if (representation == rhs.representation)
            {
                switch (representation)
                {
                    case LONG:
                        if (longValue == rhs.longValue)
                        {
                            return true;
                        }
                        break;

                    case DOUBLE:
                        if (doubleToLongBits(doubleValue) == doubleToLongBits(rhs.doubleValue))
                        {
                            return true;
                        }
                        break;
                }
            }
        }

        return false;
    }

    /**
     * Return hashCode for value. This is the underlying representations hashCode for the value
     *
     * @return int value of the hashCode
     */
    public int hashCode()
    {
        final long bits = (representation == Representation.DOUBLE) ? doubleToLongBits(doubleValue) : longValue;

        return (int)(bits ^ (bits >>> 32));
    }
}

File: main/csharp/PrimitiveValue.cs
using System;
using System.Linq;
using System.Text;

namespace Adaptive.SimpleBinaryEncoding
{
    /// <summary>
    ///     Class used to encapsulate values for primitives. Used for nullVal, minVal, maxVal, and constants
    ///     <p />
    ///     <table>
    ///         <thead>
    ///             <tr>
    ///                 <th>PrimitiveType</th>
    ///                 <th>Null</th>
    ///                 <th>Min</th>
    ///                 <th>Max</th>
    ///             </tr>
    ///         </thead>
    ///         <tbody>
    ///             <tr>
    ///                 <td>char</td>
    ///                 <td>0</td>
    ///                 <td>0x20</td>
    ///                 <td>0x7E</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>int8</td>
    ///                 <td>-128</td>
    ///                 <td>-127</td>
    ///                 <td>127</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>uint8</td>
    ///                 <td>255</td>
    ///                 <td>0</td>
    ///                 <td>254</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>int16</td>
    ///                 <td>-32768</td>
    ///                 <td>-32767</td>
    ///                 <td>32767</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>uint16</td>
    ///                 <td>65535</td>
    ///                 <td>0</td>
    ///                 <td>65534</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>int32</td>
    ///                 <td>2^31</td>
    ///                 <td>-2^31 + 1</td>
    ///                 <td>2^31 - 1</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>uint32</td>
    ///                 <td>2^32 - 1</td>
    ///                 <td>0</td>
    ///                 <td>2^32 - 2</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>int64</td>
    ///                 <td>2^63</td>
    ///                 <td>-2^63 + 1</td>
    ///                 <td>2^63 - 1</td>
    ///             </tr>
    ///             <tr>
    ///                 <td>uint64</td>
    ///                 <td>2^64 - 1</td>
    ///                 <td>0</td>
    ///                 <td>2^64 - 2</td>
    ///             </tr>
    ///         </tbody>
    ///     </table>
    /// </summary>
    public class PrimitiveValue
    {
        public enum Representation
        {
            Long,
            Double,
            ByteArray
        }

        public const long MinValueChar = 0x20;
        public const long MaxValueChar = 0x7E;
        public const long NullValueChar = 0;

        public const long MinValueInt8 = -127;
        public const long MaxValueInt8 = 127;
        public const long NullValueInt8 = -128;

        public const long MinValueUint8 = 0;
        public const long MaxValueUint8 = 254;
        public const long NullValueUint8 = 255;

        public const long MinValueInt16 = -32767;
        public const long MaxValueInt16 = 32767;
        public const long NullValueInt16 = -32768;

        public const long MinValueUint16 = 0;
        public const long MaxValueUint16 = 65534;
        public const long NullValueUint16 = 65535;

        public const long MinValueInt32 = -2147483647;
        public const long MaxValueInt32 = 2147483647;
        public const long NullValueInt32 = -2147483648;

        public const long MinValueUint32 = 0;
        public const long MaxValueUint32 = 4294967293L; // 0xFFFFFFFD
        public const long NullValueUint32 = 4294967294L; // 0xFFFFFFFE

        public const long MinValueInt64 = long.MinValue + 1; // -2^63 + 1
        public const long MaxValueInt64 = long.MaxValue; //  2^63 - 1  (SBE spec says -2^63 - 1)
        public const long NullValueInt64 = long.MinValue; // -2^63

        public const long MinValueUint64 = 0;
        public const ulong MaxValueUint64 = ulong.MaxValue - 1;
        public const ulong NullValueUint64 = ulong.MaxValue;

        public const float MinValueFloat = float.Epsilon;
        public const float MaxValueFloat = float.MaxValue;
        public const float NullValueFloat = float.NaN; // TODO: can NOT be used as a normal equality check

        public const double MinValueDouble = double.Epsilon;
        public const double MaxValueDouble = double.MaxValue;
        public const double NullValueDouble = double.NaN; // TODO: can NOT be used as a normal equality check

        private readonly byte[] _byteArrayValue;
        private readonly byte[] _byteArrayValueForLong = new byte[1];
        private readonly string _characterEncoding;
        private readonly double _doubleValue;
        private readonly long _longValue;
        private readonly Representation _representation;
        private readonly int _size;

        /// <summary>
        ///     Construct and fill in value as a long.
        /// </summary>
        /// <param name="value"> in long format </param>
        /// <param name="size"></param>
        public PrimitiveValue(long value, int size)
        {
            _representation = Representation.Long;
            _longValue = value;
            _doubleValue = 0.0;
            _byteArrayValue = null;
            _characterEncoding = null;
            _size = size;
        }

        /// <summary>
        ///     Construct and fill in value as a double.
        /// </summary>
        /// <param name="value"> in double format </param>
        /// <param name="size"></param>
        public PrimitiveValue(double value, int size)
        {
            _representation = Representation.Double;
            _longValue = 0;
            _doubleValue = value;
            _byteArrayValue = null;
            _characterEncoding = null;
            _size = size;
        }

        /// <summary>
        ///     Construct and fill in value as a byte array.
        /// </summary>
        /// <param name="value"> as a byte array </param>
        /// <param name="characterEncoding"></param>
        /// <param name="size"></param>
        public PrimitiveValue(byte[] value, string characterEncoding, int size)
        {
            _representation = Representation.ByteArray;
            _longValue = 0;
            _doubleValue = 0.0;
            _byteArrayValue = value;
            _characterEncoding = characterEncoding;
            _size = size;
        }

        /// <summary>
        ///     Return size for this PrimitiveValue for serialization purposes.
        /// </summary>
        /// <value>size for serialization</value>
        public int Size
        {
            get { return _size; }
        }

        /// <summary>
        ///     The character encoding of the byte array representation.
        /// </summary>
        /// <value>the character encoding of te byte array representation.</value>
        public string CharacterEncoding
        {
            get { return _characterEncoding; }
        }

        /// <summary>
        ///     Parse constant value string and set representation based on type
        /// </summary>
        /// <param name="value">     expressed as a String </param>
        /// <param name="primitiveType"> that this is supposed to be </param>
        /// <returns> a new <seealso cref="PrimitiveValue" /> for the value. </returns>
        public static PrimitiveValue Parse(string value, PrimitiveType primitiveType)
        {
            switch (primitiveType.Type)
            {
                case SbePrimitiveType.Char:
                    if (value.Length > 1)
                    {
                        throw new ArgumentException("Constant char value malformed: " + value);
                    }
                    return new PrimitiveValue(byte.Parse(value), 1);

                case SbePrimitiveType.Int8:
                    return new PrimitiveValue(Convert.ToInt64(value), 1);

                case SbePrimitiveType.Int16:
                    return new PrimitiveValue(Convert.ToInt64(value), 2);

                case SbePrimitiveType.Int32:
                    return new PrimitiveValue(Convert.ToInt64(value), 4);

                case SbePrimitiveType.Int64:
                    return new PrimitiveValue(Convert.ToInt64(value), 8);

                case SbePrimitiveType.UInt8:
                    return new PrimitiveValue(Convert.ToInt64(value), 1);

                case SbePrimitiveType.UInt16:
                    return new PrimitiveValue(Convert.ToInt64(value), 2);

                case SbePrimitiveType.UInt32:
                    return new PrimitiveValue(Convert.ToInt64(value), 4);

                case SbePrimitiveType.UInt64:
                    // TODO: not entirely adequate, but then again, Java doesn't have unsigned 64-bit integers...
                    // TODO to fix in .NET
                    return new PrimitiveValue(Convert.ToInt64(value), 8);

                case SbePrimitiveType.Float:
                    return new PrimitiveValue(Convert.ToDouble(value), 4);

                case SbePrimitiveType.Double:
                    return new PrimitiveValue(Convert.ToDouble(value), 8);

                default:
                    throw new ArgumentException("Unknown PrimitiveType: " + primitiveType);
            }
        }

        /// <summary>
        ///     Return long value for this PrimitiveValue
        /// </summary>
        /// <returns>value expressed as a long</returns>
        public long LongValue()
        {
            if (_representation != Representation.Long)
            {
                throw new InvalidOperationException("PrimitiveValue is not a long representation");
            }

            return _longValue;
        }

        /// <summary>
        ///     Return double value for this PrimitiveValue.
        /// </summary>
        /// <returns>value expressed as a double</returns>
        public double DoubleValue()
        {
            if (_representation != Representation.Double)
            {
                throw new InvalidOperationException("PrimitiveValue is not a double representation");
            }

            return _doubleValue;
        }

        /// <summary>
        ///     Return byte array value for this PrimitiveValue.
        /// </summary>
        /// <returns> value expressed as a byte array </returns>
        public byte[] ByteArrayValue()
        {
            if (_representation != Representation.ByteArray)
            {
                throw new InvalidOperationException("PrimitiveValue is not a byte[] representation");
            }

            return _byteArrayValue;
        }

        /// <summary>
        ///     Return byte array value for this PrimitiveValue given a particular type
        /// </summary>
        /// <param name="type"> of this value </param>
        /// <returns> value expressed as a byte array </returns>
        public byte[] ByteArrayValue(SbePrimitiveType type)
        {
            if (_representation == Representation.ByteArray)
            {
                return _byteArrayValue;
            }
            if (_representation == Representation.Long && _size == 1 && type == SbePrimitiveType.Char)
            {
                _byteArrayValueForLong[0] = (byte) _longValue;
                return _byteArrayValueForLong;
            }
            throw new InvalidOperationException("PrimitiveValue is not a byte[] representation");
        }

        /// <summary>
        ///     Return String representation of this object
        /// </summary>
        /// <returns> String representing object value </returns>
        public override string ToString()
        {
            switch (_representation)
            {
                case Representation.Long:
                    return Convert.ToString(_longValue);

                case Representation.Double:
                    return Convert.ToString(_doubleValue);

                case Representation.ByteArray:
                    return Encoding.GetEncoding(_characterEncoding).GetString(_byteArrayValue);

                default:
                    throw new InvalidOperationException("Unsupported Representation: " + _representation);
            }
        }

        /// <summary>
        ///     Determine if two values are equivalent.
        /// </summary>
        /// <param name="value"> to compare this value with </param>
        /// <returns> equivalence of values </returns>
        public override bool Equals(object value)
        {
            if (null != value && value is PrimitiveValue)
            {
                var rhs = (PrimitiveValue) value;

                if (_representation == rhs._representation)
                {
                    switch (_representation)
                    {
                        case Representation.Long:
                            return _longValue == rhs._longValue;

                        case Representation.Double:
                            return _doubleValue == rhs._doubleValue;

                        case Representation.ByteArray:
                            return _byteArrayValue.SequenceEqual(rhs._byteArrayValue);
                    }
                }
            }

            return false;
        }

        /// <summary>
        ///     Return hashCode for value. This is the underlying representations hashCode for the value
        /// </summary>
        /// <returns> int value of the hashCode </returns>
        public override int GetHashCode()
        {
            long bits;
            switch (_representation)
            {
                case Representation.Long:
                    return _longValue.GetHashCode();

                case Representation.Double:
                    return _doubleValue.GetHashCode();

                case Representation.ByteArray:
                    return _byteArrayValue.GetHashCode();

                default:
                    throw new InvalidOperationException("Unrecognised representation: " + _representation);
            }
        }
    }
}